pkg/pillar/cmd/volumemgr/create.go

```diff
@@ -63,25 +80,90 @@ func createVdiskVolume(ctx *volumemgrContext, status types.VolumeStatus,
 		return created, "", errors.New(errStr)
 	}
 
-	// create a writer for the file where we want
-	f, err := os.Create(filelocation)
-	if err != nil {
-		errStr := fmt.Sprintf("error creating target file at %s: %v", filelocation, err)
-		log.Error(errStr)
-		return created, "", errors.New(errStr)
-	}
-	defer f.Close()
+	switch persistFsType {
+	case types.PersistZFS:
+		pathToFile := ""
+		_, i, err := puller.Config(true, os.Stderr, resolver)
+		if err != nil {
+			errStr := fmt.Sprintf("error Config for ref %s: %v", ref, err)
+			log.Error(errStr)
+			return created, "", errors.New(errStr)
+		}
+		if len(i.RootFS.DiffIDs) > 0 {
+			// FIXME we expects root in the first layer for now
+			b := i.RootFS.DiffIDs[0]
+			// FIXME we need the proper way to extract file from content dir of cantainerd
+			pathToFile = filepath.Join(types.ContainerdContentDir, "blobs", b.Algorithm().String(), b.Encoded())
+		}
 
-	if _, _, err := puller.Pull(&registry.FilesTarget{Root: f, AcceptHash: true}, 0, false, os.Stderr, resolver); err != nil {
-		errStr := fmt.Sprintf("error pulling %s from containerd: %v", ref, err)
-		log.Error(errStr)
-		return created, "", errors.New(errStr)
-	}
+		if pathToFile == "" {
+			errStr := fmt.Sprintf("no blobs to convert found for ref %s", ref)
+			log.Error(errStr)
+			return created, "", errors.New(errStr)
+		}
 
-	// Do we need to expand disk?
-	if err := maybeResizeDisk(filelocation, status.MaxVolSize); err != nil {
-		log.Error(err)
-		return created, "", err
+		vSize, err := diskmetrics.GetDiskVirtualSize(log, pathToFile)
+		if err != nil {
+			errStr := fmt.Sprintf("error GetDiskVirtualSize for file %s: %v", pathToFile, err)
+			log.Error(errStr)
+			return created, "", errors.New(errStr)
+		}
+		if vSize > status.MaxVolSize {
+			log.Warnf("Virtual size (%d) of volume (%s) is larger than provided MaxVolSize (%d). "+
+				"Will use virtual size.", vSize, status.Key(), status.MaxVolSize)
+		}
+		if vSize < status.MaxVolSize {
+			vSize = status.MaxVolSize
+		}
+
+		zvolName := status.ZVolName(types.VolumeZFSPool)
+		if stdoutStderr, err := zfs.CreateDataset(log, zvolName, vSize, "on"); err != nil {
+			errStr := fmt.Sprintf("Error creating zfs zvol at %s, error=%v, output=%s",
+				zvolName, err, stdoutStderr)
+			log.Error(errStr)
+			return created, "", errors.New(errStr)
+		}
+		zVolDevice := ""
+		for i := 0; i < 5; i++ {
+			// we need repeat processing of zvols to let them time to create
```

Comment:
I am not a big fan of this "eventual consistency" loop -- I think we need to make sure we get a synchronous zvol -> /dev/zdX mapping

Reply:
I moved it to zfs.go and rework to look better, but we still need to have it due to async appearing of the device.
In this case we want to do some preparation inside the worker job and wait for result of it. Actually, we do it for QMP the same way. On one hand we do not to wait too long, say infinite (we have no progress bar in this case), on another hand it is async, so we forced to wait for it.